### PR TITLE
Optimize DAC credential selection

### PR DESCRIPTION
**Pull request by g2vinay**
Caches the working credential in chain of credentials and uses it, instead of trying the chain every time.

Closes https://github.com/Azure/azure-sdk-for-java/issues/36867

### CODE REVIEW

**Comment by g2vinay** [comment]
Pushing Unit tests next.

File: sdk/identity/azure-identity/src/main/java/com/azure/identity/ChainedTokenCredential.java [hunk 1]
```diff
@@ -56,13 +58,15 @@ public class ChainedTokenCredential implements TokenCredential {
     private static final ClientLogger LOGGER = new ClientLogger(ChainedTokenCredential.class);
     private final List<TokenCredential> credentials;
     private final String unavailableError = this.getClass().getSimpleName() + " authentication failed. ---> ";
+    private AtomicReference<TokenCredential> cachedWorkingCredential;
```
Comment (billwert) [inline sdk/identity/azure-identity/src/main/java/com/azure/identity/ChainedTokenCredential.java:61]:
```suggestion
    private AtomicReference<TokenCredential> selectedCredential;
```

maybe this instead?

Comment (billwert) [reply sdk/identity/azure-identity/src/main/java/com/azure/identity/ChainedTokenCredential.java:61]: "working" implies "non-working" somewhere which while true isn't verbiage we want maybe. :)

File: sdk/identity/azure-identity/src/main/java/com/azure/identity/ChainedTokenCredential.java [hunk 2]
```diff
@@ -79,24 +83,45 @@ public class ChainedTokenCredential implements TokenCredential {
...
+                                unavailableError + p.getClass().getSimpleName()
+                                    + " authentication failed. Error Details: " + t.getMessage(),
+                                null, t));
+                        }
+                        exceptions.add((CredentialUnavailableException) t);
+                        LOGGER.info("Azure Identity => Attempted credential {} is unavailable.",
+                            p.getClass().getSimpleName());
+                        return Mono.empty();
+                    }), 1)
+                .next();
```
Comment (billwert) [inline sdk/identity/azure-identity/src/main/java/com/azure/identity/ChainedTokenCredential.java:122]: Can we factor this into a method so we don't repeat all the logic? Could just have an extra logger in the cache case to emit an extra line that says "using cache" or the like.

Comment (g2vinay) [reply sdk/identity/azure-identity/src/main/java/com/azure/identity/ChainedTokenCredential.java:122]: Yeah, agreed.
Already have a change for it.

File: sdk/identity/azure-identity/src/main/java/com/azure/identity/ChainedTokenCredential.java (hunk 1 above)
Comment (g2vinay) [reply sdk/identity/azure-identity/src/main/java/com/azure/identity/ChainedTokenCredential.java:61]: Good point.

**Review by billwert — APPROVED** [review]

**Pull request merged**